src/app/components/Staking/FinalityProviders/FinalityProviders.tsx

```diff
@@ -9,24 +13,186 @@ import { FinalityProvider as FinalityProviderInterface } from "@/app/types/final
 import { getNetworkConfig } from "@/config/network.config";
 import { Network } from "@/utils/wallet/wallet_provider";
 
+import { FinalityProviderSearch } from "../../FinalityProviders/FinalityProviderSearch";
+import { FinalityProviderMobileSortModal } from "../../Modals/FinalityProviderMobileSortModal";
+
 import { FinalityProvider } from "./FinalityProvider";
 
 interface FinalityProvidersProps {
   finalityProviders: FinalityProviderInterface[] | undefined;
   selectedFinalityProvider: FinalityProviderInterface | undefined;
-  // called when the user selects a finality provider
   onFinalityProviderChange: (btcPkHex: string) => void;
   queryMeta: QueryMeta;
 }
 
-// Staking form finality providers
+export type SortField = "moniker" | "btcPk" | "stakeSat" | "commission";
+export type SortDirection = "asc" | "desc";
+
+const getSortValue = (
+  provider: FinalityProviderInterface,
+  field: SortField,
+) => {
+  switch (field) {
+    case "moniker":
+      return provider.description?.moniker || "";
+    case "btcPk":
+      return provider.btcPk;
+    case "stakeSat":
+      return provider.activeTVLSat;
+    case "commission":
+      return parseFloat(provider.commission);
+    default:
+      return 0;
+  }
+};
+
 export const FinalityProviders: React.FC<FinalityProvidersProps> = ({
   finalityProviders,
   selectedFinalityProvider,
   onFinalityProviderChange,
   queryMeta,
 }) => {
-  // If there are no finality providers, show loading
+  const [filteredProviders, setFilteredProviders] = useState(finalityProviders);
+  const [sortedProviders, setSortedProviders] = useState<
+    FinalityProviderInterface[]
+  >([]);
+  const [searchTerm, setSearchTerm] = useState("");
+  const [sortField, setSortField] = useState<SortField>("stakeSat");
+  const [sortDirection, setSortDirection] = useState<SortDirection>("desc");
+  const [visualSortField, setVisualSortField] = useState<SortField>("stakeSat");
+  const [visualSortDirection, setVisualSortDirection] =
+    useState<SortDirection>("desc");
+
+  const [isSortModalOpen, setIsSortModalOpen] = useState(false);
+
+  const toggleSortModal = () => setIsSortModalOpen(!isSortModalOpen);
+
+  const debouncedSearch = useMemo(
+    () =>
+      debounce((searchTerm: string) => {
+        const filtered = finalityProviders?.filter(
+          (fp) =>
+            fp.description?.moniker
+              ?.toLowerCase()
+              .includes(searchTerm.toLowerCase()) ||
+            fp.btcPk.toLowerCase().includes(searchTerm.toLowerCase()),
+        );
+        setFilteredProviders(filtered);
+      }, 300),
```

Comment:
When we extract this to the separate hook, I would suggest using `300` as constant